### ISSUE TITLE
Chunk ### can't be found

### ISSUE THREAD

**Issue by pdf** [gilbertchen/duplicacy#98]
Been testing for about a day, and have already encountered this error, which appears to completely wedge the storage for the ID containing the affected snapshot (can't backup, can't prune):

```
Chunk <hash> can't be found
```

I am using the `minios` backend.  Please let me know if there's anything I can do.

**Comment by gilbertchen** [comment]
Does minio keep a log that records file uploads and deletions?  It might be possible to find out why that chunk file is missing from the log.

This may indicate a bug in the backup or prune procedure, but it can also be caused by accidental deletions (for example, running prune with -exclusive with a backup in progress).  If the backup that gave this error is not the last backup, it means that this particular chunk had existed before (otherwise the next backup would have failed) and was deleted for some reason.

If you just want to recover the storage to a usable state, delete from the storage the corresponding snapshot file under the `snapshots/repository_id` directory, and then run this command to remove chunks that are only referenced by this backup:
```
duplicacy prune -d -exclusive -exhaustive
```
Of course, make sure no other backups are running when you run this command.

To prevent this from happening again in the future, you can:
* run a check command after each backup (which would report missing chunks)
* keep logs of every prune command (so you'll know when a chunk is deleted)

**Comment by pdf** [comment]
> Does minio keep a log that records file uploads and deletions? It might be possible to find out why that chunk file is missing from the log.

Unfortunately not by default.

> it can also be caused by accidental deletions (for example, running prune with -exclusive with a backup in progress).

Hrm, I suppose it is possible that a cron job on one host overlapped this command being run on another (only way to delete the last snapshot), since in testing I had to delete a few snapshots to get the include/exclude lists right.

> If the backup that gave this error is not the last backup

It was the only backup for this id.

> - run a check command after each backup (which would report missing chunks)
> - keep logs of every prune command (so you'll know when a chunk is deleted)

I'll add these to my cron jobs for the time being, hopefully this was a once-off caused by an exclusive prune.  However, not having any way to safely delete a recent snapshot is somewhat problematic, though I understand that may be a difficult task.

**Comment by gilbertchen** [comment]
Try to avoid the -exclusive option as much as possible.  If you don't provide this option to the prune command, the two-step fossil collection algorithm will kick in which is guaranteed to work even with ongoing backups.  It may not free up the space immediately, but eventually it will.

**Comment by pdf** [comment]
Sure, the CLI won't allow pruning of the last backup for an id without exclusive though.

**Comment by gilbertchen** [comment]
Right, I think there should be a different option that allows the last backup to be deleted without assuming the exclusive access.

**Comment by pdf** [comment]
Closed in favour of #104
